src/kernel-install/ukify.py

```diff
@@ -0,0 +1,233 @@
+import itertools
+import pathlib
+import os
+import shlex
+import subprocess
+import tempfile
+import typing
+
+efi_arches = {
+        # host_arch: efi_arch
+        'x86'    : 'ia32',
+        'x86_64' : 'x64',
+        'arm'    : 'arm',
+        'aarch64': 'aa64',
+        'riscv64': 'riscv64',
+}
+
+def guess_efi_arch():
+    arch = os.uname().machine
+    efi_arch = efi_arches[arch]
+
+    if arch == 'x86_64':
+        try:
+            size = open('/sys/firmware/efi/fw_platform_size').read().strip()
+        except FileNotFoundError:
+            pass
+        else:
+            if int(size) == 32:
+                efi_arch = 'ia32'
+
+    print(f'Host arch {arch!r}, efi arch {efi_arch!r}')
+    return efi_arch
+
+def shell_join(cmd):
+    return ' '.join(shlex.quote(str(x)) for x in cmd)
+
+@dataclasses.dataclass
+class Section:
+    name: str
+    content: str
+    offset: int
+    tmpfile: typing.Optional[typing.IO] = None
+
+    @classmethod
+    def create(cls, name, content, offset):
+        if isinstance(content, str):
+            tmp = tempfile.NamedTemporaryFile(mode='wt', prefix=f'tmp{name}')
+            tmp.write(content)
+            tmp.flush()
+            content = tmp.name
+        else:
+            tmp = None
+
+        return cls(name, content, offset, tmpfile=tmp)
+
+    @classmethod
+    def parse_arg(cls, s):
+        parts = s.split(':')
+        if len(parts) < 2 or len(parts) > 3:
+            raise ValueError(f'Cannot parse section spec: {s!r}')
+        name, content = parts[:2]
+        offset = int(parts[2], base=0) if len(parts) >= 3 else None
+        return cls.create(name, content, offset)
+
+
+def parse_args():
+    p = argparse.ArgumentParser(
+        description='Build and sign Unified Kernel Images',
+        usage='''\
+usage: ukify [options…] linux initrd
+       ukify -h | --help
+''')
+
+    # Suppress printing of usage synopsis on errors
+    p.error = lambda message: p.exit(2, f'{p.prog}: error: {message}\n')
+
+    p.add_argument('linux',
+                   type=pathlib.Path)
+    p.add_argument('initrd',
+                   type=pathlib.Path)
+
+    p.add_argument('--cmdline')
+
+    p.add_argument('--os-release',
+                   type=pathlib.Path)
+
+    p.add_argument('--devicetree',
+                   type=pathlib.Path)
+    p.add_argument('--splash',
+                   type=pathlib.Path)
+
+    p.add_argument('--efi-arch',
+                   choices=('ia32', 'x64', 'arm', 'aa64', 'riscv64'))
+
+    p.add_argument('--stub',
+                   type=pathlib.Path)
+
+    p.add_argument('--section',
+                   dest='sections',
+                   metavar='NAME:CONTENTS[:OFFSET]',
+                   type=Section.parse_arg,
+                   action='append')
+
+    p.add_argument('--output', '-o',
+                   nargs='?')
+
+    p.add_argument('--key',
+                   type=pathlib.Path)
+    p.add_argument('--cert',
+                   type=pathlib.Path)
```

Comment:
If we do want to hardcode `sbsign` usage then we need to be more clever here. `sbsign` supports `--engine` which allows you to specify an arbitrary string for `--key`. 

It would allow us to utilize PKCS11 to sign with a TPM and/or hardware keys.

Preferably I'd prefer if we leave this out of the helper, but that is personal and not very strong opinion.

Reply:
I think it's nice to do signing, at least optionally. In particular, I think it's nice to call `systemd-measure` too, because we have all the sections specified here. And then it seems reasonable to have all the parts in the same tool.

I'll do the following: add `--sign-engine` and pass that through as `--engine` to sbsign, along with the key and cert args.

Reply:
so i think right from the beginning we should clarify in the option names what the key/cert is used for. i.e. ultimately we probably want at least three sets of keys involved:

1. one key for SB signing the UKI
2. one key for PCR signing the UKI for the initrd boot phase
3. one key for PCR signing the UKI for all boot phases from initrd until shutdown begins.

Yes, we need distinct keys for the latter: people must enroll their LUKS volumes or creds with one of the two keys depending on whether they want it to be accessible after initrd transition or only before.

Reply:
I have `--pcr-private-key`, `--pcr-public-key`, `--sb-engine`, `--sb-key`, `--sb-cert`. ("sb" instead of "secure-boot" because this abbreviation should be pretty clear in this contents and the commandline is annoying long already.)

Reply:
i see no reason to shorten the options needlessly. This is not an interface people use daily, but a build tool. I think expressivity is a good thing here. i.e. immediatelly grokking what this command line does is more important than to be able to type it fast. Hence "secureboot" is a much more common syntax than "sb".

Also, "--pcr-private-key=" vs."--sb-key="? Why does one get the "private" and the other doesn't?

I'd always qualify with "private" given we are talking about key pairs, and we should make sure we never confuse things between the public and private of the two...

Also as mentioned I am pretty sure we should have two key pairs for the PCR case.

Reply:
> * one key for PCR signing the UKI for the initrd boot phase
> 
> * one key for PCR signing the UKI for all boot phases from initrd until shutdown begins.

Hmm, but `systemd-measure sign` doesn't seem to support phases at all? I'm all confused now. Shouldn't the output consist of multiple combos of `pkfp` + `pol` + `sig`?

Reply:
hmm, somwhere my commit for measure.c got lost that added that. fuck. must be a rebase fuckup.

There was supposed to be a hookup in verb_sign() much like in verb_calculate() for measure_phase(). But apaprently that got lost. Yikes.

Reply:
is that something to fix for 252?

Reply:
#25202

Reply:
I prepped #25224 now, to allow generating multiple signatures by multiple different keys.